unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -352,6 +351,196 @@ public int indexOf(UTF8String v, int start) {
+    }
+    return i;
+  }
+
+  // Locate to the last position in byte for a given code point
+  private int lastByteIndex(int codePoint) {
+    int i = firstByteIndex(codePoint);
+    return i + numBytesForFirstByte(getByte(i)) - 1;
+  }
+
+  /**
+   * Returns the index within this string of the last occurrence of the
+   * specified substring, searching backward starting at the specified index.
+   * @param v the substring to search for.
+   * @param startCodePoint the index to start search from
+   * @return the index of the last occurrence of the specified substring,
+   *         searching backward from the specified index,
+   *         or {@code -1} if there is no such occurrence.
+   */
+  public int lastIndexOf(UTF8String v, int startCodePoint) {
+    return lastIndexOf(v, v.numChars(), startCodePoint);
+  }
+  public int lastIndexOf(UTF8String v, int vNumChars, int startCodePoint) {
+    if (v.numBytes == 0) {
+      return 0;
+    }
+    if (numBytes == 0) {
+      return -1;
+    }
+    int fromIndexEnd = lastByteIndex(startCodePoint);
+    do {
+      if (fromIndexEnd - v.numBytes + 1 < 0 ) {
+        return -1;
+      }
+      if (ByteArrayMethods.arrayEquals(
+          base, offset + fromIndexEnd - v.numBytes + 1, v.base, v.offset, v.numBytes)) {
+        int count = 0; // count from right most to the match end in byte.
+        while (fromIndexEnd >= 0) {
+          count++;
+          fromIndexEnd = firstOfCurrentCodePoint(fromIndexEnd) - 1;
+        }
+        return count - vNumChars;
+      }
+      fromIndexEnd  = firstOfCurrentCodePoint(fromIndexEnd) - 1;
+    } while (fromIndexEnd >= 0);
+    return -1;
+  }
+
+  /**
+   * Finds the n-th last index within a String.
+   * This method uses {@link String#lastIndexOf(String)}.</p>
+   *
+   * @param str  the String to check, may be null
+   * @param searchStr  the String to find, may be null
+   * @param ordinal  the n-th last <code>searchStr</code> to find
+   * @return the n-th last index of the search String,
+   *  <code>-1</code> if no match or <code>null</code> string input
+   */
+  public static int lastOrdinalIndexOf(
+          UTF8String str,
+          UTF8String searchStr,
+          int ordinal) {
+    if (str == null || searchStr == null) {
+      return -1;
+    }
+    return doOrdinalIndexOf(str, searchStr, searchStr.numChars(), ordinal, true);
+  }
+
+  /**
+   * Finds the n-th index within a String, handling <code>null</code>.
+   * A <code>null</code> String will return <code>-1</code>
+   *
+   * @param str  the String to check, may be null
+   * @param searchStr  the String to find, may be null
+   * @param ordinal  the n-th <code>searchStr</code> to find
+   * @return the n-th index of the search String,
+   *  <code>-1</code> if no match or <code>null</code> string input
+   */
+  public static int ordinalIndexOf(
+          UTF8String str,
+          UTF8String searchStr,
+          int ordinal) {
+    if (str == null || searchStr == null) {
```

Comment:
Usually we don't do null value checking, as the caller will guarantee the validity. For example: `java.lang.String.indexOf(String str)`.